aspnetcore/fundamentals/host/platform-specific-configuration/samples/2.x/HostingStartupApp/HostingStartupApp.csproj

```diff
@@ -0,0 +1,36 @@
+<Project Sdk="Microsoft.NET.Sdk.Web">
+
+  <PropertyGroup>
+    <TargetFramework>netcoreapp2.1</TargetFramework>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.AspNetCore.App" Version="2.1.2" />
+    <PackageReference Include="HostingStartupPackage" Version="1.0.0" />
+  </ItemGroup>
+
+  <!--
+    The following PropertyGroup sets the package restore sources for the app.
+
+    The HostingStartupPackage bin/Debug folder is added.
+    
+    This permits the app to restore the HostingStartupPackage from its 
+    project's published output. There is no need to publish the package to 
+    nuget.org.
+  -->
+  <PropertyGroup>
+    <RestoreSources>$(RestoreSources);https://api.nuget.org/v3/index.json;../HostingStartupPackage/bin/Debug</RestoreSources>
```

Comment:
Isn't this typically defined in a _NuGet.config_ file? What's the advantage of putting it here?

Reply:
Two things here:
1. pointing to a bin/Debug output as restore source is a bit hacky but I don't know the context here. (what if it fails, what if NuGet decides that you haven't changed any feeds/references and does a no-op (incremental) restore etc.)

2. There's also `RestoreAdditionalProjectSources` for project-specific sources. But now idea how well a "manage package for solution" dialog works for this.

The property-based approach is nice if you don't want to explain NuGet.config and the mechanism behind and it is easy to set in `Directory.Build.props` for all projects, even for local paths by using `$(MSBuildThisFileDirectory)local\nupkgs` from it. Also, you can make use of standard MSBuild conditions to alter the restore sources which is why dotnet/aspnet moved to them to be able to use different sources for different types of builds which would require editing NuGet.config for every build otherwise.

Reply:
@dasMulli 

True ... it's a bit of a hack. It has the following benefits/goals in this context:

1. It allows the dev to create and use the *HostingStartupPackage* NuGet package in the *HostingStartupApp* by just building that *HostingStartupPackage* NuGet package project and then building/running the *HostingStartupApp*.
1. They don't need to move anything (i.e., they don't need to move the NuGet package out of the *HostingStartupPackage* project). They don't need to take any additional steps in *HostingStartupApp* to find and consume the package.
1. They don't need to put the package on nuget.org. **_That's the really important goal here._**

> you haven't changed any feeds/references and does a no-op (incremental) restore

I thought that if the NuGet package is rebuilt that NuGet will pick up on that (without a package version change). Does NuGet go purely by the package's version on incremental restore? That is a problem if that's the case. I would like to have this set up so that the dev can make a change to the NuGet package project and have it picked up on restore/run of the *HostingStartupApp*.

> RestoreAdditionalProjectSources 

Can the *additional project source* be a NuGet package that is then consumed by the app? If I can use `RestoreAdditionalProjectSources` pointed to the NuGet package project, that would be great ... but is that going to work for consuming a NuGet package? If so, do I then drop the `<RestoreSources>` property?

Reply:
@guardrex AFAIK you'd have to do a `dotnet nuget locals all --clear` if you change package contents without changing the version anyway.
That's why you should always use project (P2P) references during development or use a local "global packages cache" to nuke in between full rebuilds (CI runs).

Reply:
Why not use a project reference?

Reply:
Thanks @dasMulli ... I pushed a commit to make sure devs know that if they change the package project to clear caches. That looks good.

I'm going to leave the `<RestoreSources>` in the app as is. Unless @pakrym says 'don't do that,' it seems like a very clean, easy-to-use, easy-to-grok setup for the demo. I'm just trying to show how the hosting startup consumer takes *a real package*, and I don't want the dev to have to put a hosting startup package on nuget.org or need to sweat a *Nuget.config* file with the source (which I think buries how the source is being set).

Your advice on the caches should go well :+1:. They just need to do that each time they make changes to the hosting startup package. 🤞

https://github.com/aspnet/Docs/pull/7678/commits/37562f03d2d3d4a5c1d8ace2bae44ee5af38918c